joyd.1
.\" This file Copyright 2000 Christian Garbs <mitch@uni.de>
.\" 
.\" It may be distributed under the GNU Public License, version 2, or
.\" any higher version.  See section COPYING of the GNU Public license
.\" for conditions under which this file may be redistributed.
.\"
.\" As I don't have a clue of the Nroff format, this document is based
.\" on the top(1) man page which is Copyright 1992 Robert J. Nation 
.\" (nation@rocket.sanders.lockheed.com) and was afterwards modified
.\" on 1994/04/25 by Michael Shields <mjshield@nyx.cs.du.edu> and on
.\" 1996/01/27 by Helmut Geyer.
.
.de It
.TP 0.5i
.B "\\$1 "
..
.TH JOYD 1 "Apr 13 2000" "v0.0.7"
.SH NAME
.B joyd
\- a joystick daemon
.SH SYNOPSIS
.B joyd
.RB [ \-v | configfile ]
.
.SH DESCRIPTION
.B joyd
is a daemon that allows you to trigger commands on your computer
by using your joystick. It can execute any program.
.PP
You can bind commands to single buttons, movements of your stick or
combinations. There can be as much commands as you like, you can even
bind multiple commands to one combination.
.PP
.B joyd
uses very few resources: when nothing happens at the joystick
port, it just sleeps, needing no CPU time.
.PP
When run in daemon mode,
.B joyd
switches to the background. It can itself still be controlled by
joystick commands or signals (e.g. SIGHUP).
.B joyd
is also able to log its output to syslog. 
.PP
.SH "COMMAND\-LINE OPTIONS"
.It -v
.B joyd
will print its version number and exit.
.It configfile
This is where
.B joyd
will look for a configuration file. If no configuration file is given,
.B joyd
will use the compiled in default.
.PP
.
.SH RUNNING JOYD
When
.B joyd
is started, it will either run in normal mode or in daemon
(background) mode. This can be set within the configuration file. If 
you want to stop a normal
.B joyd
simply press
.B ctrl-c
.PP
If you want to stop a
.B joyd
daemon, you have to look up its PID using
.B ps aux
and then kill it by typing
.B kill <PID>
If you have only one
.B joyd
running, you could also say
.B killall joyd
(you are able to start multiple instances of
.B joyd
- they can even share the same joystick device).
Note that you can also bind the exit command to your joystick.

When
.B joyd
is running, you can execute the commands given in the configuration
file by using your joystick.
.B joyd
scans the axes and buttons of your joystick. Normally axis 1 is
left-right and axis 2 is up-down. If your joystick provides more axes
then you can also use them.

The joystick buttons can be scanned in two modes: normal and
shift. The normal mode executes a command after one or more buttons
have been pressed and released. A shift button does not execute a
command but can modify the behaviour of a normal key. A shift button
does not have to be released. Look at the configuration file
.B joydrc.shift
for an example of this.

When joyd receives the SIGHUP signal, it will reread its configuration
file (for example
.B kill -HUP <PID>
where PID can be found out using
.B ps aux
).
.
.SH FILES
.I joydrc 
The default configuration file.
.
.SH "SEE ALSO"
.BR joydrc(5)
.
.\"SH
.\"BUGS
.\".
.SH AUTHOR
.B joyd
was written by Christian Garbs <mitch@uni.de>. Please send bug reports
to him.


HISTORY
joyd 0.2.2 2000-10-11
~~~~~~~~~~~~~~~~~~~~~
- a call without any command line arguments now successfully
  reads the compiled in default configuration file
- fixed some typos in the documentation



joyd 0.2.1 2000-05-25
~~~~~~~~~~~~~~~~~~~~~
 - fixed permissions in tar.gz-archive (everything was 0777)
 - config.stdout renamed to config.std_out due to an error in 
   /usr/include/stdio.h in older libc6 systems



joyd 0.2.0 2000-04-22
~~~~~~~~~~~~~~~~~~~~~
 - support of GNU autoconf/configure
 - documentation is now by default installed under /usr/local/doc/joyd
 - added a configuration example for control sequences (joydrc.sequence)
 - "joyd --version" will now also print version number (like "joyd -v")



joyd 0.0.7 2000-04-13
~~~~~~~~~~~~~~~~~~~~~
 - removed type cast of fgetc(), casting EOF to char instead
 - changed some C++ style comment lines to C style comments



joyd 0.0.6 2000-04-06
~~~~~~~~~~~~~~~~~~~~~
 - merged joyreadbutton.c and joyreadaxis.c into joyread.c
 - type cast of fgetc() to signed char because of compiler error
   on PPC systems
 - program version displayed on startup still was "0.0.3"
 - a new entry to the joystick configuration file (CONFIG)



joyd 0.0.5 2000-03-19
~~~~~~~~~~~~~~~~~~~~~
 - fixed a problem with joysticks that report buttons as axes with
   values of 1 and -1
 - added joyreadbutton and joyreadaxis to the joyd package
 - first two entries in the joystick configuration file (CONFIG)



joyd 0.0.4 2000-02-09
~~~~~~~~~~~~~~~~~~~~~
 - again some minor changes to the documentation, i.e. the default
   values are now listed in the joydrc(5) manpage



joyd 0.0.3 2000-01-17
~~~~~~~~~~~~~~~~~~~~~
 - calibration values are all wrong: the range is from -32768 to
   32767, not from -16384 to 16383
 - minor changes in documentation
 - added version information to the header files



joyd 0.0.2 2000-01-10
~~~~~~~~~~~~~~~~~~~~~
 - the global variable config was not declared correctly 
 - manual pages were added
 - "joyd -v" now prints the version number
 - config.c and config.o have been renamed to options.c and options.h
   because config.h files are generated by autoconfigure
 - the ActionLoop() function got split up into smaller parts



joyd 0.0.1 2000-01-09
~~~~~~~~~~~~~~~~~~~~~
first version
 - added signal handlers
 - added syslog routines
 - added background mode


src/joyd/joyd.c
/*
 *    joyd.c
 *
 *    this is the main file of:
 *
 *    joyd 0.2.2   ---   The Joystick Daemon
 *
 *    2000 (C) by Christian Garbs <mitch@uni.de>
 */

/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

/*
 * You can execute commands by using your joystick, joypad or whatever
 * is supported by the Linux Joystick Driver. This allows you to
 * reboot your computer when your keyboard is hung or you can even use
 * your system without any keyboard at all. 
 *
 * I use one of my computers (no screen, no keyboard, no hard disks)
 * for listening to mp3-files - with this program I can stop, start
 * and skip songs, change the volume, shutdown the computer and much
 * more just by using an joypad (it's in fact an original SNES
 * joypad!).
 *
 * If you use this and find it useful, please let me know!
 * If you find a bug or have any suggestions or questions: tell me!
 *
 * My mail adress is:    <mitch@uni.de>
 * My homepage is here:  http://www.cgarbs.de/
 *
 *
 *                                    Now enjoy this.
 *                                    Christian */

/*
 * This program is based on jstest.c Version 1.2 by Vojtech Pavlik,
 * which "is also intended to serve as an example implementation for
 * those who wish to learn how to write their own joystick using
 * applications."
 *
 * That's just what I did. That's GPL. Thanks!
 */


/*****************************************************************************
 *                                                                           *
 *     "Even the little things make me happy. It's that... that ordinary     *
 *      life I want back... I want back."                (Tsukino Usagi)     *
 *                                                                           *
 *****************************************************************************/


/*     These are just the changes to this file.
 *     See the HISTORY file for changes on the whole package.
 */


/*  joyd 0.0.1 2000-01-09
 *
 *  first version
 *  - added signal handlers
 *  - added syslog routines
 *  - added background mode
 */

/*  joyd 0.0.2 2000-01-10
 *
 *  - the global variable config is now declared here and declared
 *    external in joyd.h
 *  - "joyd -v" prints the version number
 */

/*  joyd 0.0.3 2000-01-17
 *
 *  - changed the JOYD_VERSION macro to PROGRAM_VERSION
 */

/*  joyd 0.0.4 2000-02-09
 *  joyd 0.0.5 2000-03-19
 *  joyd 0.0.6 2000-04-06
 *  joyd 0.0.7 2000-04-13
 *
 *  (no changes)
 */

/*  joyd 0.2.0 2000-04-22
 *
 *  - "joyd --version" prints the version number (like "joyd -v")
 *  - print of version number additionally prints short help text
 */

/*  joyd 0.2.1 2000-05-25
 *
 *  (no changes)
 */

/*  joyd 0.2.2 2000-10-11
 *
 *  - a call without any command line arguments now successfully
 *    reads the compiled in default configuration file
 */

/*
 *  2do:
 *  ? supply Sys V start/stop script (for /etc/rc.d/)
 *  ? implement shift_axes like shift_keys
 */

/*****************************************************************************
 * now goes it loose...                                                      *
 *****************************************************************************/


/*
 * system include files
 */

#include <stdio.h>
#include <stdlib.h>

/*
 * my headers and definitions
 */

#include "joyd.h"
#include "daemon.h"
#include "joystick.h"
#include "log.h"
#include "options.h"
#include "signal.h"
#include "string.h"

/*
 * the global variable
 */

TCONFIG config;

/*
 * now to the functions
 */



int main (int argc, char **argv)
/* simple main program */
{
	/* don't tell me that this condition looks ugly... */
	if (    (argc > 2)
		|| (
			(argc == 2)
			&& (
				(strcmp(argv[1],PRINT_VERSION)      == 0)
				|| (strcmp(argv[1],PRINT_VERSION_LONG) == 0)
				|| (strcmp(argv[1],PRINT_HELP_LONG)    == 0)
				)
			)
		) {
		printf(PROGRAM_VERSION "\n");
		printf(PROGRAM_INFOTEXT "\n");
		exit(0);
	};

	SetDefaultValues();

	if (config.debug > 1) {
		Print(stdout,PROGRAM_VERSION," starting up");
	}

	InstallSignalHandler();

	if (argc > 1) {
		free(config.config_file);
		config.config_file = AllocString(argv[1]);
	}
	
	ReadConfigFile();

	if (config.daemon) {
		ForkToBackground();
	}
	
	OpenJoystick();

	AddMissingCalibrations();

	ActionLoop();

	CloseJoystick();

	Print(stdout,PROGRAM_VERSION," quits now (should've never come around here!?)");

	return 0;
}
